src/hp/derp/client_conn.rs

```diff
@@ -118,60 +115,52 @@ impl ClientConnManager {
     /// the client
     /// Call `shutdown` to close the read and write loops before dropping the ClientConnManager
     #[allow(clippy::too_many_arguments)]
-    pub fn new<R, W, P>(
+    pub fn new<P>(
         key: PublicKey,
         conn_num: usize,
-        reader: R,
-        writer: W,
+        io: MaybeTlsStream,
         can_mesh: bool,
         write_timeout: Option<Duration>,
         channel_capacity: usize,
-        server_channel: mpsc::Sender<ServerMessage<R, W, P>>,
+        server_channel: mpsc::Sender<ServerMessage<P>>,
     ) -> ClientConnManager
     where
-        R: AsyncRead + Unpin + Send + Sync + 'static,
-        W: AsyncWrite + Unpin + Send + Sync + 'static,
         P: PacketForwarder,
     {
         let done = CancellationToken::new();
         let client_id = (key.clone(), conn_num);
         let (send_queue_s, send_queue_r) = mpsc::channel(channel_capacity);
 
         let (disco_send_queue_s, disco_send_queue_r) = mpsc::channel(channel_capacity);
-        let (send_pong_s, send_pong_r) = mpsc::channel(channel_capacity);
         let (peer_gone_s, peer_gone_r) = mpsc::channel(channel_capacity);
         let (mesh_update_s, mesh_update_r) = mpsc::channel(channel_capacity);
-        let conn_writer = ClientConnWriter {
+
+        let preferred = Arc::from(AtomicBool::from(false));
+
+        let conn_io = ClientConnIo {
             can_mesh,
-            writer,
+            io,
             timeout: write_timeout,
             send_queue: send_queue_r,
             disco_send_queue: disco_send_queue_r,
-            send_pong: send_pong_r,
             peer_gone: peer_gone_r,
             mesh_update_r,
             mesh_update_s: mesh_update_s.clone(),
-        };
 
-        let preferred = Arc::from(AtomicBool::from(false));
-        let conn_reader = ClientConnReader {
-            can_mesh,
-            reader,
             key: key.clone(),
-            send_pong: send_pong_s,
-            server_channel: server_channel.clone(),
             preferred: Arc::clone(&preferred),
+            server_channel: server_channel.clone(),
         };
 
-        // start writer loop
-        let writer_done = done.clone();
-        let writer_client_id = client_id.clone();
-        let writer_server_channel = server_channel.clone();
-        let writer_handle = tokio::spawn(async move {
-            let key = writer_client_id.0;
-            let conn_num = writer_client_id.1;
-            let res = conn_writer.run(writer_done).await;
-            let _ = writer_server_channel
+        // start io loop
+        let io_done = done.clone();
+        let io_client_id = client_id.clone();
+        let io_server_channel = server_channel.clone();
+        let io_handle = tokio::task::spawn(async move {
```

Comment:
you have some `trace!` calls in `conn_io.run()`, would be great to attach `.instrument(tracing::debug_span!("conn_io"))` or similar to this async block as that would give nice context for those.  Like at https://github.com/n0-computer/iroh/blob/6df66f479aac274f8231aaabbe3d11f07521b6a4/src/hp/netcheck.rs#LL1702C21-L1702C21